File: src/main/java/com/matag/game/cardinstance/modifiers/Counters.java
package com.matag.game.cardinstance.modifiers;

import static java.util.stream.Collectors.toList;

import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.matag.cards.ability.type.AbilityType;
import com.matag.game.cardinstance.ability.CardInstanceAbility;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@ToString
@EqualsAndHashCode
@JsonAutoDetect(
  fieldVisibility = JsonAutoDetect.Visibility.NONE,
  setterVisibility = JsonAutoDetect.Visibility.NONE,
  getterVisibility = JsonAutoDetect.Visibility.NONE,
  isGetterVisibility = JsonAutoDetect.Visibility.NONE,
  creatorVisibility = JsonAutoDetect.Visibility.NONE
)
@Getter
public class Counters {
  @JsonProperty
  private int plus1Counters;
  @JsonProperty
  private int minus1Counters;
  @JsonProperty
  private List<AbilityType> keywordCounters = new ArrayList<>();

  public void addPlus1Counters(int n) {
    plus1Counters += n;
    balanceCounters();
  }

  public void addMinus1Counters(int n) {
    minus1Counters += n;
    balanceCounters();
  }

  public void addKeywordCounter(AbilityType abilityType) {
    keywordCounters.add(abilityType);
  }

  public List<CardInstanceAbility> getKeywordCountersAbilities() {
    return keywordCounters.stream()
      .map(CardInstanceAbility::new)
      .map(CardInstanceAbility::getCardInstanceAbility)
      .collect(toList());
  }

  private void balanceCounters() {
    var min = Math.min(plus1Counters, minus1Counters);
    plus1Counters -= min;
    minus1Counters -= min;
  }
}

File: src/main/kotlin/com/matag/game/cardinstance/modifiers/Counters.kt
package com.matag.game.cardinstance.modifiers

import com.matag.cards.ability.Ability
import com.matag.cards.ability.type.AbilityType
import kotlin.math.min

data class Counters(
    var plus1Counters: Int = 0,
    var minus1Counters: Int = 0,
    var keywordCounters: List<AbilityType> = listOf()
) {

    fun addPlus1Counters(n: Int) {
        plus1Counters += n
        balanceCounters()
    }

    fun addMinus1Counters(n: Int) {
        minus1Counters += n
        balanceCounters()
    }

    fun addKeywordCounter(abilityType: AbilityType) {
        keywordCounters = keywordCounters + abilityType
    }

    val keywordCountersAbilities: List<Ability>
        get() = keywordCounters
            .map { Ability(it) }

    private fun balanceCounters() {
        val min = min(plus1Counters, minus1Counters)
        plus1Counters -= min
        minus1Counters -= min
    }
}
